File: src/test/java/org/springframework/samples/petclinic/e2e/pages/WelcomePage.java
/*
 * Copyright 2012-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.springframework.samples.petclinic.e2e.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

/**
 * Page Object for the PetClinic Welcome/Home page
 */
public class WelcomePage extends BasePage {

	public WelcomePage(Page page) {
		super(page);
	}

	public void navigate() {
		navigateTo("/");
	}

	public Locator getWelcomeImage() {
		return page.locator("img[src*='pets.png']");
	}

	public Locator getFindOwnersLink() {
		return page.locator("a[href*='/owners/find']");
	}

	public Locator getVeterinariansLink() {
		return page.locator("a[href*='/vets.html']");
	}

	public void clickFindOwners() {
		getFindOwnersLink().click();
	}

	public void clickVeterinarians() {
		getVeterinariansLink().click();
	}

	public boolean isWelcomePageDisplayed() {
		return getWelcomeImage().isVisible() && getTitle().contains("PetClinic");
	}

}

File: e2e-tests/pages/WelcomePage.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage';

export class WelcomePage extends BasePage {
  
  constructor(page: Page) {
    super(page);
  }

  // Page-specific elements
  get welcomeHeading(): Locator {
    return this.page.locator('h2');
  }

  get petImage(): Locator {
    return this.page.locator('img[src*="pets.png"]');
  }

  // Page actions
  async goto(): Promise<void> {
    await this.page.goto('/');
    await this.page.waitForLoadState('networkidle');
  }

  // Page assertions
  async assertWelcomePageElements(): Promise<void> {
    await expect(this.welcomeHeading).toBeVisible();
    await expect(this.welcomeHeading).toContainText('Welcome');
    await expect(this.petImage).toBeVisible();
  }

  async assertPageLoaded(): Promise<void> {
    await this.assertPageTitle(/PetClinic/);
    await this.assertWelcomePageElements();
    await this.assertNavigation();
  }
}